### ISSUE TITLE
radare2 - 'wa' failed to assemble x64 opcode

### ISSUE THREAD

**Issue by nixawk** [radare/radare2#8251]
```
Lab     : Ubuntu 17.04 x64
Kernel  : Linux lab 4.10.0-32-generic #36-Ubuntu SMP Tue Aug 8 12:10:06 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
```

```
code@lab:~/debug$ radare2 -A -a x86 -b 64 -d function_params
Process with PID 26841 started...
= attach 26841 26841
bin.baddr 0x55ab490d5000
USING 55ab490d5000
Assuming filepath /home/code/debug/function_params
asm.bits 64
[x] Analyze all flags starting with sym. and entry0 (aa)
[Cannot determine xref search boundariesr references (aar)
[x] Analyze len bytes of instructions for references (aar)
[Oops invalid rangen calls (aac)
[x] Analyze function calls (aac)
[ ] [*] Use -AA or aaaa to perform additional experimental analysis.
[x] Constructing a function name for fcn.* and sym.func.* functions (aan))
[0x7fe8ae6afc20]> afl
0x55ab490d5000    8 409  -> 412  sym.imp.__cxa_finalize
0x55ab490d54f8    3 23           sym._init
0x55ab490d5520    1 16           sub.__cxa_finalize_248_520
0x55ab490d5530    1 43           entry0
0x55ab490d5560    4 50   -> 44   sym.deregister_tm_clones
0x55ab490d55a0    4 66   -> 57   sym.register_tm_clones
0x55ab490d55f0    5 50           sym.__do_global_dtors_aux
0x55ab490d5630    4 48   -> 42   sym.frame_dummy
0x55ab490d5660    1 61           sym.add
0x55ab490d569d    1 48           sym.main
0x55ab490d56d0    4 101          sym.__libc_csu_init
0x55ab490d5740    1 2            sym.__libc_csu_fini
0x55ab490d5744    1 9            sym._fini
0x55ab492d5fd8    1 1020         reloc.__cxa_finalize_216
[0x7fe8ae6afc20]> s sym.main
[0x55ab490d569d]> pdf
            ;-- main:
/ (fcn) sym.main 48
|   sym.main ();
|           ; DATA XREF from 0x55ab490d554d (entry0)
|           0x55ab490d569d      55             push rbp
|           0x55ab490d569e      4889e5         mov rbp, rsp
|           0x55ab490d56a1      41b906000000   mov r9d, 6
|           0x55ab490d56a7      41b805000000   mov r8d, 5
|           0x55ab490d56ad      b904000000     mov ecx, 4
|           0x55ab490d56b2      ba03000000     mov edx, 3
|           0x55ab490d56b7      be02000000     mov esi, 2
|           0x55ab490d56bc      bf01000000     mov edi, 1
|           0x55ab490d56c1      e89affffff     call sym.add
|           0x55ab490d56c6      b800000000     mov eax, 0
|           0x55ab490d56cb      5d             pop rbp
\           0x55ab490d56cc      c3             ret
```

----

## Issue details

```
[0x56375e9a769d]> wa?
|Usage: wa[of*] [arg]
| wa nop           write nopcode using asm.arch and asm.bits
| wa* mov eax, 33  show 'wx' op with hexpair bytes of assembled opcode
| "wa nop;nop"     assemble more than one instruction (note the quotes)
| waffoo.asm       assemble file and write bytes
| wao?             show help for assembler operation on current opcode (hack)
```

```
[0x55ab490d569d]> "wa push rbp;mov rbp,rsp;mov r9d, 6"
Cannot assemble 'mov r9d, 6' at line 9
```

## function_params.c

```
// $: gcc -g -o function_params function_params.c

int
add(int a, int b, int c, int d, int e, int f)
{
    int x = 1;
    return a + b + c + d + e + f;
}


int
main()
{
    add(1, 2, 3, 4, 5, 6);
    return 0;
}
```

**Comment by Maijin** [comment]
Greetings,

x86.nz (x86 handmade assembler) is one of the many x86 assemblers available in radare2 and is updated frequently.  **Please add a test case so we can fix x86.nz in https://github.com/radare/radare2-regressions/blob/master/t.asm/x86/nz/x86_asm** You can do so with github web editor without even cloning the repo.

You can contribute to x86.nz by completing the following file https://github.com/radare/radare2/blob/master/libr/asm/p/asm_x86_nz.c.
You can also use keystone(http://keystone-engine.org) within radare2/rasm2 by installing the radare2 plugin via r2pm : 

```
	r2pm -i keystone-lib
	r2pm -i keystone
then
	rasm2 -a x86.ks…
or in radare2 session
	e asm.assembler = x86.ks
```

Others x86 assemblers are also available trough r2pm or master (see rasm2 -L list):

```
a___  16 32 64   x86.as      LGPL3   Intel X86 GNU Assembler
a___  16 32 64   x86.nasm    LGPL3   X86 nasm assembler
a___  16 32 64   x86.nz      LGPL3   x86 handmade assembler
ad__  32         x86.olly    GPL2    OllyDBG X86 disassembler
```

**Comment by Maijin** [comment]
Hello,

Ensure you are using radare2 from git, if you're unsure paste output of `r2 -v` here.
To install radare2 from git, first uninstall your version of radare2 and clean your distro. Then use `git clone https://github.com/radare/radare2 && cd radare2 && ./sys/install.sh`, verify your version and check if there is no error using `r2 -v`.

**Comment by nixawk** [comment]
Thanks @Maijin. radare2 is updated and compiled again, as follow:

```
Lab     : Ubuntu 17.04 x64
Kernel  : Linux lab 4.10.0-32-generic #36-Ubuntu SMP Tue Aug 8 12:10:06 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
```

```
$ git pull https://github.com/radare/radare2/
$ cd radare2
$ sudo ./sys/install.sh

$ radare2 -v
radare2 1.7.0-git 15640 @ linux-x86-64 git.1.6.0-401-gb645c7fd1
commit: b645c7fd1ade3f2b85b0bd255795ef53862292b1 build: 2017-08-22__09:39:22
```

A new issue here,  r2 fails to disassebmle opcode.

```
code@lab:~/debug$ r2 -d function_params
Process with PID 5470 started...
= attach 5470 5470
bin.baddr 0x558a7aba5000
Using 0x558a7aba5000
asm.bits 64
 -- This page intentionally left blank.
[0x7f47aff79c20]> aaa
[x] Analyze all flags starting with sym. and entry0 (aa)
TODO: esil-vm not initialized
[Cannot determine xref search boundariesr references (aar)
[x] Analyze len bytes of instructions for references (aar)
[x] Analyze function calls (aac)
[x] Use -AA or aaaa to perform additional experimental analysis.
[x] Constructing a function name for fcn.* and sym.func.* functions (aan)
ptrace (PT_ATTACH): No such process
= attach 3 3
[0x7f47aff79c20]> afl
0x558a7aba54f8    3 23           sym._init
0x558a7aba5520    1 16           sym.imp.__cxa_finalize
0x558a7aba5530    1 43           entry0
0x558a7aba5560    4 50   -> 44   sym.deregister_tm_clones
0x558a7aba55a0    4 66   -> 57   sym.register_tm_clones
0x558a7aba55f0    5 50           sym.__do_global_dtors_aux
0x558a7aba5630    4 48   -> 42   sym.frame_dummy
0x558a7aba5660    1 61           sym.add
0x558a7aba569d    1 48           sym.main
0x558a7aba56d0    4 101          sym.__libc_csu_init
0x558a7aba5740    1 2            sym.__libc_csu_fini
0x558a7aba5744    1 9            sym._fini
0x558a7ada5ad2    1 790          sym.imp.__libc_start_main
0x558a7ada5fd8    1 1020         reloc.__libc_start_main_216
[0x7f47aff79c20]> s sym.main 
[0x558a7aba569d]> pdf
            ;-- main:
/ (fcn) sym.main 48
|   sym.main ();
|              ; DATA XREF from 0x558a7aba554d (entry0)
|           0x558a7aba569d      ff             invalid
|           0x558a7aba569e      ff             invalid
|           0x558a7aba569f      ff             invalid
|           0x558a7aba56a0      ff             invalid
|           0x558a7aba56a1      ff             invalid
|           0x558a7aba56a2      ff             invalid
|           0x558a7aba56a3      ff             invalid
|           0x558a7aba56a4      ff             invalid
|           0x558a7aba56a5      ff             invalid
|           0x558a7aba56a6      ff             invalid
|           0x558a7aba56a7      ff             invalid
|           0x558a7aba56a8      ff             invalid
|           0x558a7aba56a9      ff             invalid

```

**Comment by radare** [comment]
Never ever do aaa on debugger and never ever run it in the dynamic linker because i doubt u want to analyze the linker

**Comment by nixawk** [comment]
If apt installs radare2, the disassembler works good.

```
code@lab:~/debug$ sudo apt-get -y install radare2
code@lab:~/debug$ radare2 -v 
radare2 1.1.0 9999999 @ linux-x86-64 git.1.1.0
commit: HEAD build: 2017-03-03
```

----

```
code@lab:~/debug$ radare2 -d function_params
Process with PID 82002 started...
= attach 82002 82002
bin.baddr 0x5651ee345000
USING 5651ee345000
Assuming filepath /home/code/debug/function_params
asm.bits 64
```

```
[0x7f8bc4b8fc20]> aaa;afl
[x] Analyze all flags starting with sym. and entry0 (aa)
[Cannot determine xref search boundariesr references (aar)
[x] Analyze len bytes of instructions for references (aar)
[Oops invalid rangen calls (aac)
[x] Analyze function calls (aac)
[ ] [*] Use -AA or aaaa to perform additional experimental analysis.
[x] Constructing a function name for fcn.* and sym.func.* functions (aan))
0x5651ee345000    8 409  -> 412  sym.imp.__cxa_finalize
0x5651ee3454f8    3 23           sym._init
0x5651ee345520    1 16           sub.__cxa_finalize_248_520
0x5651ee345530    1 43           entry0
0x5651ee345560    4 50   -> 44   sym.deregister_tm_clones
0x5651ee3455a0    4 66   -> 57   sym.register_tm_clones
0x5651ee3455f0    5 50           sym.__do_global_dtors_aux
0x5651ee345630    4 48   -> 42   sym.frame_dummy
0x5651ee345660    1 61           sym.add
0x5651ee34569d    1 48           sym.main
0x5651ee3456d0    4 101          sym.__libc_csu_init
0x5651ee345740    1 2            sym.__libc_csu_fini
0x5651ee345744    1 9            sym._fini
0x5651ee545fd8    1 1020         reloc.__cxa_finalize_216
```

```
[0x7f8bc4b8fc20]> pdf @ sym.main 
            ;-- main:
/ (fcn) sym.main 48
|   sym.main ();
|           ; DATA XREF from 0x5651ee34554d (entry0)
|           0x5651ee34569d      55             push rbp
|           0x5651ee34569e      4889e5         mov rbp, rsp
|           0x5651ee3456a1      41b906000000   mov r9d, 6
|           0x5651ee3456a7      41b805000000   mov r8d, 5
|           0x5651ee3456ad      b904000000     mov ecx, 4
|           0x5651ee3456b2      ba03000000     mov edx, 3
|           0x5651ee3456b7      be02000000     mov esi, 2
|           0x5651ee3456bc      bf01000000     mov edi, 1
|           0x5651ee3456c1      e89affffff     call sym.add
|           0x5651ee3456c6      b800000000     mov eax, 0
|           0x5651ee3456cb      5d             pop rbp
\           0x5651ee3456cc      c3             ret
```

**Comment by Svenito** [comment]
Closed via #8265 8265
